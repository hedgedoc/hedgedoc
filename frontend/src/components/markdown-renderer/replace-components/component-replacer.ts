/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { Element, Node } from 'domhandler'
import { isText } from 'domhandler'
import type { ReactElement } from 'react'

export type ValidReactDomElement = ReactElement | string | null

export type SubNodeTransform = (node: Node, subKey: number | string) => ValidReactDomElement

export type NativeRenderer = () => ValidReactDomElement

export const DO_NOT_REPLACE = Symbol()

export type NodeReplacement = ValidReactDomElement | typeof DO_NOT_REPLACE

export enum ReplacerPriority {
  LOWER = 1,
  NORMAL = 0,
  HIGHER = -1
}

/**
 * Base class for all component replacers.
 * Component replacers detect structures in the HTML DOM from markdown it
 * and replace them with some special react components.
 */
export abstract class ComponentReplacer {
  /**
   * Extracts the content of the first text child node.
   *
   * @param node the node with the text node child
   * @return the string content
   */
  protected static extractTextChildContent(node: Element): string {
    const childrenTextNode = node.children[0]
    return isText(childrenTextNode) ? childrenTextNode.data : ''
  }

  /**
   * Applies the given {@link SubNodeTransform sub node transformer} to every children of the given {@link Node}.
   *
   * @param node The node whose children should be transformed
   * @param subNodeTransform The transformer that should be used.
   * @return The children as react elements.
   */
  protected static transformChildren(node: Element, subNodeTransform: SubNodeTransform): ValidReactDomElement[] {
    return node.children.map((value, index) => subNodeTransform(value, index))
  }

  /**
   * Should be used to reset the replacers internal state before rendering.
   */
  public reset(): void {
    // left blank for overrides
  }

  /**
   * Checks if the current node should be altered or replaced and does if needed.
   *
   * @param node The current html dom node
   * @param subNodeTransform should be used to convert child elements of the current node
   * @param nativeRenderer renders the current node without any replacement
   * @return the replacement for the current node or undefined if the current replacer replacer hasn't done anything.
   */
  public abstract replace(
    node: Element,
    subNodeTransform: SubNodeTransform,
    nativeRenderer: NativeRenderer
  ): NodeReplacement

  /**
   * Defines that a replacer should be preferred more or less than other replacers.
   *
   * @return the replacer priority that gets compared to others.
   */
  public getPriority(): ReplacerPriority {
    return ReplacerPriority.NORMAL
  }
}
