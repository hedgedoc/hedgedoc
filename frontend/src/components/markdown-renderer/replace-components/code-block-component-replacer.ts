/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NodeReplacement } from './component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from './component-replacer'
import type { Element } from 'domhandler'
import type { FunctionComponent } from 'react'
import React from 'react'

export interface CodeProps {
  code: string
}

/**
 * Checks if the given checked node is a code block with a specific language attribute and creates an react-element that receives the code.
 */
export class CodeBlockComponentReplacer extends ComponentReplacer {
  constructor(
    private component: FunctionComponent<CodeProps>,
    private language: string
  ) {
    super()
  }

  replace(node: Element): NodeReplacement {
    const code = CodeBlockComponentReplacer.extractTextFromCodeNode(node, this.language)
    return code ? React.createElement(this.component, { code: code }) : DO_NOT_REPLACE
  }

  /**
   * Extracts the text content if the given {@link Element} is a code block with a specific language.
   *
   * @param element The {@link Element} to check.
   * @param language The language that code block should be assigned to.
   * @return The text content or undefined if the element isn't a code block or has the wrong language attribute.
   */
  public static extractTextFromCodeNode(element: Element, language: string): string | undefined {
    return element.name === 'code' && element.attribs['data-highlight-language'] === language && element.children[0]
      ? ComponentReplacer.extractTextChildContent(element)
      : undefined
  }
}
