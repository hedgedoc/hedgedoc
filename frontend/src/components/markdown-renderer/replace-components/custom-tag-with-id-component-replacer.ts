/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NodeReplacement } from './component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from './component-replacer'
import type { Element } from 'domhandler'
import type { FunctionComponent } from 'react'
import React from 'react'

export interface IdProps {
  id: string
}

/**
 * Replaces custom tags that have just an id (<app-something id="something"/>) with react elements.
 */
export class CustomTagWithIdComponentReplacer extends ComponentReplacer {
  constructor(
    private component: FunctionComponent<IdProps>,
    private tagName: string
  ) {
    super()
  }

  public replace(node: Element): NodeReplacement {
    const id = this.extractId(node)
    return id ? React.createElement(this.component, { id: id }) : DO_NOT_REPLACE
  }

  /**
   * Checks if the given {@link Element} is a custom tag and extracts its `id` attribute.
   *
   * @param element The element to check.
   * @return the extracted id or undefined if the element isn't a custom tag or has no id attribute.
   */
  private extractId(element: Element): string | undefined {
    return element.name === this.tagName && element.attribs && element.attribs.id ? element.attribs.id : undefined
  }
}
