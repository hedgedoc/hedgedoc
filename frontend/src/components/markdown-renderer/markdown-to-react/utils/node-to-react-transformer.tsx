/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { LinemarkerMarkdownExtension } from '../../extensions/linemarker/linemarker-markdown-extension'
import type { LineWithId } from '../../extensions/linemarker/types'
import type {
  ComponentReplacer,
  NodeReplacement,
  ValidReactDomElement
} from '../../replace-components/component-replacer'
import { DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import { convertNodeToReactElement } from '@hedgedoc/html-to-react/dist/convertNodeToReactElement'
import { Optional } from '@mrdrogdrog/optional'
import type { Element, Node } from 'domhandler'
import { isTag } from 'domhandler'
import React from 'react'

type LineIndexPair = [startLineIndex: number, endLineIndex: number]

/**
 * Converts {@link Node domhandler nodes} to react elements by using direct translation or {@link ComponentReplacer replacers}.
 */
export class NodeToReactTransformer {
  private lineIds: LineWithId[] = []
  private replacers: ComponentReplacer[] = []

  public setLineIds(lineIds: LineWithId[]): void {
    this.lineIds = lineIds
  }

  public setReplacers(replacers: ComponentReplacer[]): void {
    this.replacers = new Array(...replacers).sort(this.compareReplacers.bind(this))
  }

  private compareReplacers(replacerA: ComponentReplacer, replacerB: ComponentReplacer): number {
    const priorityA = replacerA.getPriority()
    const priorityB = replacerB.getPriority()
    if (priorityA === priorityB) {
      return 0
    } else if (priorityA < priorityB) {
      return -1
    } else {
      return 1
    }
  }

  /**
   * Resets all replacers before rendering.
   */
  public resetReplacers(): void {
    this.replacers.forEach((replacer) => replacer.reset())
  }

  /**
   * Converts the given {@link Node} to a React element.
   *
   * @param node The {@link Node DOM node} that should be translated.
   * @param index The index of the node within its parents child list.
   * @return the created react element
   */
  public translateNodeToReactElement(node: Node, index: number | string): ValidReactDomElement {
    return isTag(node)
      ? this.translateElementToReactElement(node, index)
      : convertNodeToReactElement(node, index, this.translateNodeToReactElement.bind(this))
  }

  /**
   * Translates the given {@link Element} to a React element.
   *
   * @param element The {@link Element DOM element} that should be translated.
   * @param index The index of the element within its parents child list.
   * @return the created react element
   */
  private translateElementToReactElement(element: Element, index: number | string): ValidReactDomElement {
    const elementKey = this.calculateUniqueKey(element).orElseGet(() => (-index).toString())
    const replacement = this.findElementReplacement(element, elementKey)
    if (replacement === null) {
      return null
    } else if (replacement === DO_NOT_REPLACE) {
      return this.renderNativeNode(element, elementKey)
    } else if (typeof replacement === 'string') {
      return replacement
    } else {
      return React.cloneElement(replacement, {
        ...(replacement.props as Record<string, unknown>),
        key: elementKey
      })
    }
  }

  /**
   * Calculates the unique key for the given {@link Element}.
   *
   * @param element The element for which the unique key should be calculated.
   * @return An {@link Optional} that contains the unique key or is empty if no key could be found.
   */
  private calculateUniqueKey(element: Element): Optional<string> {
    if (!element.attribs) {
      return Optional.empty()
    }

    return Optional.ofNullable(element.prev)
      .map((lineMarker) => NodeToReactTransformer.extractLineIndexFromLineMarker(lineMarker))
      .map(([startLineIndex, endLineIndex]) =>
        NodeToReactTransformer.convertMarkdownItLineIndexesToInternalLineIndexes(startLineIndex, endLineIndex)
      )
      .flatMap((adjustedLineIndexes) => this.findLineIdsByIndex(adjustedLineIndexes))
      .map(([startLine, endLine]) => `${startLine.id}_${endLine.id}`)
  }

  /**
   * Asks every saved replacer if the given {@link Element element} should be
   * replaced with another react element or not.
   *
   * @param element The {@link Element} that should be checked.
   * @param elementKey The unique key for the element
   * @return The replacement or {@link DO_NOT_REPLACE} if the element shouldn't be replaced with a custom component.
   */
  private findElementReplacement(element: Element, elementKey: string): NodeReplacement {
    const transformer = this.translateNodeToReactElement.bind(this)
    const nativeRenderer = () => this.renderNativeNode(element, elementKey)
    for (const componentReplacer of this.replacers) {
      const replacement = componentReplacer.replace(element, transformer, nativeRenderer)
      if (replacement !== DO_NOT_REPLACE) {
        return replacement
      }
    }
    return DO_NOT_REPLACE
  }

  /**
   * Extracts the start and end line indexes that are saved in a line marker element
   * and describe in which line, in the Markdown code, the node before the marker ends
   * and which the node after the marker starts.
   *
   * @param lineMarker The line marker that saves a start and end line index.
   * @return the extracted line indexes
   */
  private static extractLineIndexFromLineMarker(lineMarker: Node): LineIndexPair | undefined {
    if (!isTag(lineMarker) || lineMarker.tagName !== LinemarkerMarkdownExtension.tagName || !lineMarker.attribs) {
      return
    }
    const startLineInMarkdown = lineMarker.attribs['data-start-line']
    const endLineInMarkdown = lineMarker.attribs['data-end-line']
    if (startLineInMarkdown === undefined || endLineInMarkdown === undefined) {
      return
    }

    const startLineIndex = Number(startLineInMarkdown)
    const endLineIndex = Number(endLineInMarkdown)

    if (isNaN(startLineIndex) || isNaN(endLineIndex)) {
      return
    }

    return [startLineIndex, endLineIndex]
  }

  /**
   * Converts markdown-it line indexes to internal line indexes.
   * The differences are:
   *   - Markdown it starts to count at 1, but we start at 0
   *   - Line indexes in markdown-it are start(inclusive) to end(exclusive). But we need start(inclusive) to end(inclusive).
   *
   * @param startLineIndex The start line index from markdown it
   * @param endLineIndex The end line index from markdown it
   * @return The adjusted start and end line index
   */
  private static convertMarkdownItLineIndexesToInternalLineIndexes(
    startLineIndex: number,
    endLineIndex: number
  ): LineIndexPair {
    return [startLineIndex - 1, endLineIndex - 2]
  }

  /**
   * Renders the given node without any replacement
   *
   * @param node The node to render
   * @param key The unique key for the node
   * @return The rendered {@link ValidReactDomElement}
   */
  private renderNativeNode = (node: Element, key: string): ValidReactDomElement => {
    if (node.attribs === undefined) {
      node.attribs = {}
    }

    return convertNodeToReactElement(node, key, this.translateNodeToReactElement.bind(this))
  }

  private findLineIdsByIndex([startLineIndex, endLineIndex]: LineIndexPair): Optional<[LineWithId, LineWithId]> {
    const startLine = this.lineIds[startLineIndex]
    const endLine = this.lineIds[endLineIndex]
    return startLine === undefined || endLine === undefined ? Optional.empty() : Optional.of([startLine, endLine])
  }
}
