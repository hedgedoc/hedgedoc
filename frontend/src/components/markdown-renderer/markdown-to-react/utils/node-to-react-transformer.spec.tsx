/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NodeReplacement } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE, ReplacerPriority } from '../../replace-components/component-replacer'
import { NodeToReactTransformer } from './node-to-react-transformer'
import { Element } from 'domhandler'
import type { ReactElement, ReactHTMLElement } from 'react'

describe('node to react transformer', () => {
  let nodeToReactTransformer: NodeToReactTransformer
  let defaultTestSpanElement: Element

  beforeEach(() => {
    defaultTestSpanElement = new Element('span', { 'data-test': 'test' })
    nodeToReactTransformer = new NodeToReactTransformer()
  })

  describe('replacement', () => {
    it('can translate an element without any replacer', () => {
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement
      expect(translation.type).toEqual('span')
      expect(translation.props).toEqual({ children: [], 'data-test': 'test' })
    })

    it('can replace an element nothing', () => {
      nodeToReactTransformer.setReplacers([
        new (class extends ComponentReplacer {
          replace(): NodeReplacement {
            return null
          }
        })()
      ])
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement
      expect(translation).toEqual(null)
    })

    it('will prioritize a high priority replacer over a normal priority replacer', () => {
      nodeToReactTransformer.setReplacers([
        new (class extends ComponentReplacer {
          getPriority(): ReplacerPriority {
            return ReplacerPriority.NORMAL
          }

          replace(): NodeReplacement {
            return <span>Replacer O</span>
          }
        })(),
        new (class extends ComponentReplacer {
          getPriority(): ReplacerPriority {
            return ReplacerPriority.HIGHER
          }

          replace(): NodeReplacement {
            return <span>Replacer X</span>
          }
        })()
      ])
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement
      expect(translation).toMatchSnapshot()
    })

    it('will prioritize a normal priority replacer over a low priority replacer', () => {
      nodeToReactTransformer.setReplacers([
        new (class extends ComponentReplacer {
          getPriority(): ReplacerPriority {
            return ReplacerPriority.LOWER
          }

          replace(): NodeReplacement {
            return <span>Replacer M</span>
          }
        })(),
        new (class extends ComponentReplacer {
          getPriority(): ReplacerPriority {
            return ReplacerPriority.NORMAL
          }

          replace(): NodeReplacement {
            return <span>Replacer Y</span>
          }
        })()
      ])
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement
      expect(translation).toMatchSnapshot()
    })

    it('can translate an element with no matching replacer', () => {
      nodeToReactTransformer.setReplacers([
        new (class extends ComponentReplacer {
          replace(): NodeReplacement {
            return DO_NOT_REPLACE
          }
        })()
      ])
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement

      expect(translation.type).toEqual('span')
      expect(translation.props).toEqual({ children: [], 'data-test': 'test' })
    })

    it('can replace an element', () => {
      nodeToReactTransformer.setReplacers([
        new (class extends ComponentReplacer {
          replace(): NodeReplacement {
            return <div data-test2={'test2'} />
          }
        })()
      ])
      const translation = nodeToReactTransformer.translateNodeToReactElement(defaultTestSpanElement, 1) as ReactElement

      expect(translation.type).toEqual('div')
      expect(translation.props).toEqual({ 'data-test2': 'test2' })
    })
  })

  describe('key calculation', () => {
    beforeEach(() => {
      nodeToReactTransformer.setLineIds([
        {
          id: 1,
          line: 'test'
        }
      ])
    })

    it('can calculate a fallback key', () => {
      const result = nodeToReactTransformer.translateNodeToReactElement(
        defaultTestSpanElement,
        1
      ) as ReactHTMLElement<HTMLDivElement>

      expect(result.type).toEqual('span')
      expect(result.key).toEqual('-1')
    })

    it('can calculate a key based on line markers and line keys', () => {
      const lineMarker = new Element('app-linemarker', { 'data-start-line': '1', 'data-end-line': '2' })
      defaultTestSpanElement.prev = lineMarker
      const rootElement: Element = new Element('div', {}, [lineMarker, defaultTestSpanElement])

      const result = nodeToReactTransformer.translateNodeToReactElement(
        rootElement,
        1
      ) as ReactHTMLElement<HTMLDivElement>
      const resultSpanTag = (result.props.children as ReactElement[])[1]

      expect(result.type).toEqual('div')
      expect(resultSpanTag.type).toEqual('span')
      expect(resultSpanTag.key).toEqual('1_1')
    })
  })
})
