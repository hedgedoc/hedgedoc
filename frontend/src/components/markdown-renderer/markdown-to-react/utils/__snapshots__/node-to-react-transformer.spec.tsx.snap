// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node to react transformer replacement will prioritize a high priority replacer over a normal priority replacer 1`] = `
<span>
  Replacer X
</span>
`;

exports[`node to react transformer replacement will prioritize a normal priority replacer over a low priority replacer 1`] = `
<span>
  Replacer Y
</span>
`;
